fix: allow `flox pull` to overwrite environment with `--force` flag

charmitro opened the pull request:
## Proposed Changes

When using `flox pull --force` while in an environment, it fails to overwrite the environment. This commit allows overwriting it by checking if the `force` flag is set.

Closes #1205 

## Release Notes

Allow overwriting the environment when pulling with `--force`, i.e. `flox pull --force`.

mkenigs commented on cli/tests/environment-pull.bats:
```diff
@@ -297,6 +297,17 @@ function add_insecure_package() {
+# pulling an environment without packages for the current platform
+@test "pull environment inside the same environment with '--force' flag" {
+  update_dummy_env "owner" "name"
+  make_incompatible "owner" "name"
```
`make_incompatible` is only needed when testing how we handle incompatible environments specifically
```suggestion
```

charmitro replied:
Ack & resolved.

charmitro commented:
> We have a couple of ways to run the test suite:
> 
>     * `just test-all`: Runs the whole thing, unit tests for `flox` and `pkgdb`, integration tests for `flox` and `pkgdb`, and the "functional" tests
> 
>     * `just test-cli`: Runs just the unit tests and integration tests for `flox`
> 
>     * `just integ-tests`: integration tests for `flox`
> 
>     * `just unit-tests`: unit tests for `flox`
> 
>     * `just test-pkgdb`: unit and integration tests for `pkgdb`
> 
>     * Push to GitHub, then hope and pray that CI passes (kidding, kind of)
> 
> 
> When in doubt, use `just test-all`

Ack. Thank you! Running `just integ-tests` shows that my test-case as passing, I'm more concerned if my test-case is correct.

mkenigs commented on cli/flox/src/commands/environment.rs:
```diff
@@ -1696,7 +1696,7 @@ impl Pull {
         message: &str,
     ) -> Result<()> {
-        if dot_flox_path.exists() {
+        if dot_flox_path.exists() && !force {
```
suggestion: I think we should make sure the old environment is cleaned up - there could be other files in `.flox`, and there could be other files living in FLOX_CACHE_DIR. I think cleaning up would look like trying to `open()` the `.flox` directory. If that succeeds, call `Environment::delete()`. If it fails, just remove the directory (or make sure somehow that there aren't any files leftover that we don't overwrite in the rest of the method

charmitro replied:
> suggestion: I think we should make sure the old environment is cleaned up - there could be other files in `.flox`, and there could be other files living in FLOX_CACHE_DIR. I think cleaning up would look like trying to `open()` the `.flox` directory. If that succeeds, call `Environment::delete()`. If it fails, just remove the directory (or make sure somehow that there aren't any files leftover that we don't overwrite in the rest of the method)

Good suggestion, thanks! I ended up using `ManagedEnvironment::open(...)`. I'm not 100% percent sure about it though.

mkenigs replied:
I led you astray sorry it should actually be `open_path()` - that calls `DotFlox::open()` and then `into_concrete_environment()` which could end up calling either `ManagedEnvironment::open()` or `PathEnvironment::open()`. `open()` was probably not specific enough 😳 

Basically a `.flox` directory could either contain an environment created with `flox init`, or it could contain one created by `flox pull`. This should be able to handle both of them.

charmitro replied:
Before committing it, how's this?

```rust
        if dot_flox_path.exists() {                                                                                                                                                                                                                            
            if force {                                                                                                                                                                                                                                         
                match open_path(flox, &dot_flox_path) {                                                                                                                                                                                                        
                    Ok(concrete_env) => match concrete_env {                                                                                                                                                                                                   
                        ConcreteEnvironment::Path(env) => {                                                                                                                                                                                                    
                            env.delete(flox)                                                                                                                                                                                                                   
                                .context("Failed to delete existing environment")?;                                                                                                                                                                            
                        },                                                                                                                                                                                                                                     
                        ConcreteEnvironment::Managed(env) => {                                                                                                                                                                                                 
                            env.delete(flox)                                                                                                                                                                                                                   
                                .context("Failed to delete existing environment")?;                                                                                                                                                                            
                        },                                                                                                                                                                                                                                     
                        ConcreteEnvironment::Remote(_) => { },                                                                                                                                                                                                                                     
                    },                                                                                                                                                                                                                                         
                    Err(_) => {                                                                                                                                                                                                                                
                        fs::remove_dir_all(&dot_flox_path).context(format!(                                                                                                                                                                                    
                            "Failed to remove existing .flox directory at {:?}",                                                                                                                                                                               
                            dot_flox_path                                                                                                                                                                                                                      
                        ))?;                                                                                                                                                                                                                                   
                    },                                                                                                                                                                                                                                         
                }                                                                                                                                                                                                                                              
            } else {                                                                                                                                                                                                                                           
                bail!(                                                                                                                                                                                                                                         
                    "An environment already exists at {:?}. Use --force to overwrite.",                                                                                                                                                                        
                    dot_flox_path                                                                                                                                                                                                                              
                );                                                                                                                                                                                                                                             
            }                                                                                                                                                                                                                                                  
        }                                                                                                                                                                                                                                                      
```

mkenigs replied:
Looks great!

charmitro replied:
Thanks, committed!

mkenigs reviewed (changes requested):
Thanks for the contribution! Looks like its headed in the right, and thanks for adding the test!

charmitro commented:
It seems the CI can't find my branch because the origin branch is in a fork?

mkenigs commented:
> It seems the CI can't find my branch because the origin branch is in a fork?

Yeah, we have to push your commits to a branch in our repo and then we can trigger it. Just ping if you ever want us to run it. I was thinking I'd trigger it after you addressed review?

ghudgins commented:
also don't forget to add yourself to CONTRIBUTORS.csv to accept our CLA. See https://github.com/flox/flox/blob/main/CONTRIBUTING.md#cla

charmitro commented:
> Yeah, we have to push your commits to a branch in our repo and then we can trigger it. Just ping if you ever want us to run it. I was thinking I'd trigger it after you addressed review?

Yes, it would be less noisy if we just trigger after the resolving the review.

> also don't forget to add yourself to CONTRIBUTORS.csv to accept our CLA. See https://github.com/flox/flox/blob/main/CONTRIBUTING.md#cla

Ack. Thanks!

mkenigs reviewed (changes requested):
See thread about using `open_path`

mkenigs reviewed (approved):
Thanks so much!

tomberek commented:
Woooooo!!!!
